src/iosrv/sched.c

```diff
@@ -26,73 +26,953 @@
+
+	D_ASSERT(policy_ops[sched_policy].process_io != NULL);
+	policy_ops[sched_policy].process_io(dx);
+}
+
+static inline bool
+should_enqueue_req(struct dss_xstream *dx, struct sched_req_attr *attr)
+{
+	struct sched_info	*info = &dx->dx_sched_info;
+
+	if (sched_prio_disabled || info->si_stop)
+		return false;
+
+	D_ASSERT(attr->sra_type == SCHED_REQ_GC ||
+		 attr->sra_type == SCHED_REQ_IO ||
+		 attr->sra_type == SCHED_REQ_MIGRATE);
+
+	/* For VOS xstream only */
+	return dx->dx_main_xs;
+}
+
+static void
+req_enqueue(struct dss_xstream *dx, struct sched_request *req)
+{
+	struct sched_info	*info = &dx->dx_sched_info;
+	struct sched_req_attr	*attr = &req->sr_attr;
+	struct sched_req_info	*sri;
+
+	D_ASSERT(req->sr_pool_info != NULL);
+	sri = &req->sr_pool_info->spi_req_array[attr->sra_type];
+
+	D_ASSERT(d_list_empty(&req->sr_link));
+	if (attr->sra_type == SCHED_REQ_IO) {
+		D_ASSERT(policy_ops[sched_policy].enqueue_io != NULL);
+		policy_ops[sched_policy].enqueue_io(dx, req, NULL);
+	} else {
+		d_list_add_tail(&req->sr_link, &sri->sri_req_list);
+	}
+	req->sr_enqueue_ts = info->si_cur_ts;
+
+	sri->sri_req_cnt++;
+	info->si_req_cnt++;
+}
+
+int
+sched_req_enqueue(struct dss_xstream *dx, struct sched_req_attr *attr,
+		  void (*func)(void *), void *arg)
+{
+	struct sched_request	*req;
+
+	if (!should_enqueue_req(dx, attr))
+		return req_kickoff_internal(dx, attr, func, arg);
+
+	D_ASSERT(attr->sra_type < SCHED_REQ_MAX);
+	req = req_get(dx, attr, func, arg, ABT_THREAD_NULL);
+	if (req == NULL) {
+		D_ERROR("XS(%d): get req failed.\n", dx->dx_xs_id);
+		return -DER_NOMEM;
+	}
+	req_enqueue(dx, req);
+
 	return 0;
 }
 
+void
+sched_req_yield(struct sched_request *req)
+{
+	struct dss_xstream	*dx = dss_current_xstream();
+
+	D_ASSERT(req != NULL);
+	if (!should_enqueue_req(dx, &req->sr_attr)) {
+		ABT_thread_yield();
+		return;
+	}
+
+	D_ASSERT(req->sr_ult != ABT_THREAD_NULL);
+	req_enqueue(dx, req);
+
+	ABT_self_suspend();
+}
+
+void
+sched_req_sleep(struct sched_request *req, uint32_t msecs)
+{
+	struct dss_xstream	*dx = dss_current_xstream();
+	struct sched_info	*info = &dx->dx_sched_info;
+	struct sched_request	*tmp;
+	int			 rc;
+
+	D_ASSERT(req != NULL);
+	if (msecs == 0 || info->si_stop || req->sr_abort) {
+		sched_req_yield(req);
+		return;
+	}
+
+	/* Don't put GC & aggregation ULTs in sleep under space pressure */
+	if (req->sr_attr.sra_type == SCHED_REQ_GC) {
+		D_ASSERT(req->sr_pool_info != NULL);
+		rc = check_space_pressure(dx, req->sr_pool_info, false);
+		if (rc != 0) {	/* under space pressure */
+			sched_req_yield(req);
+			return;
+		}
+	}
+
+	D_ASSERT(req->sr_ult != ABT_THREAD_NULL);
+	req->sr_wakeup_time = info->si_cur_ts + msecs;
+
+	D_ASSERT(d_list_empty(&req->sr_link));
+	/* Sleep list is sorted in wakeup time ascending order */
+	d_list_for_each_entry_reverse(tmp, &info->si_sleep_list, sr_link) {
+		if (req->sr_wakeup_time < tmp->sr_wakeup_time)
+			continue;
+		d_list_add(&req->sr_link, &tmp->sr_link);
```

Comment:
Oops, we should stop loop here. This nasty bug can cause memory corruption when there are multiple sleepers.